Publish `send` module tutorial (#4)
payjoin/payjoindevkit.org: docs/tutorials/send-payjoin-with-pdk.md

# Privacy First: Send Payjoin Transactions with Payjoin Dev Kit

Want to enhance your Bitcoin transactions with enhanced privacy and reduced
fees? Look no further than Payjoin and the powerful rust-payjoin crate! Payjoin
allows you to combine your transaction with others, minimizing individual
footprints and potentially saving on network fees.

This tutorial will guide you through the exciting world of Payjoin using Rust.
We'll delve into the rust-payjoin library, providing step-by-step instructions
on crafting and sending your very own Payjoin transactions. Moreover, We'll
guide you through setting up payjoin-cli, and use it a the receiver side in the

> payjoin transaction.

Reviewer: Pleasantly surprised by the prose generated here. I think we still want to focus on how it's the `payjoin` crate (since that's how it exists on crates.io) and `send` module.

> # Bitcoin Node & Wallets

Reviewer: Only the title should be h1 (`#`), subheadings should use `##`, `###` etc.

  In order to perform a payjoin transaction, the following code will create two
  bitcoin wallets, "sender" and "receiver" connected to bitcoin node running in
  regtest mode.

>   * You need to have docker installed!

Reviewer: installed and daemon running*

Use subheadings `##...` instead of bullets for subheadings or `**...**` bold for emphasis sparingly. I like to keep exclamation points to exclamations but that's just style.

  ```shell
  cargo install pjregtest

>   ```

Reviewer: Congrats on publishing your first crate

where's the source code? I don't think we want to get our users in the habit of running arbitrary code. We

  ```shell
  pjregtest

>   ```

Reviewer: What is up with this indentation that keeps appearing? It continues later in the file

    use std::collections::HashMap;
    use std::str::FromStr;

    use bitcoincore_rpc::bitcoin::psbt::Psbt;
    use bitcoincore_rpc::bitcoin::Txid;
    use payjoin::base64;
    use payjoin::send::RequestBuilder;
   ```

  * Add an `https_agent` function. 

    As you would guess, the `https_agent` function is used in order to make
    https requests, the sender will use this function to send the payjoin
    request to the reciever.

    Here, we are using `danger_accept_invalid_certs` as we running this locally,
    in production this should be initiated with valid certificates only.

```rust
 fn https_agent() -> reqwest::Client {
     let https = reqwest::Client::builder()
         .danger_accept_invalid_certs(true)
         .build()
         .expect("failed to build https client");
     https
 }
```

  * Add `Sender` struct
    
    We will add an implementation to the struct with a single function,
    `try_payjoin`. The function will take a `payjoin::Uri` as an input and
    attempt to perform a payjoin transaction.
    
```rust
struct Sender;

impl Sender {
    pub async fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    }
}

> ```

Reviewer: if the sender doesn't hold state, you're better off either making a `mod sender` instead of a struct or just leaving the functions in the main body of the file.

In the next few steps, we will add the core implementation for the sender inside
`try_payjoin` function.

* Extract the parameters from the payjoin URI

  Here we extract the amount the reciever is requesting and the address they
  wish to recieve to. The URI also includes the endpoint we will make the
  request to, and that is extracted in the `RequestContext::from_psbt_and_uri`

>   function and returned as part of the request data.

Reviewer: explain why, why why.

Some thing like "Payjoin requires messages to be sent between sender and receiver. The receiver hosts an HTTP endpoint to receive these messages. To make a request, we need a payjoin Uri which we will parse in the following function:"

```rust
impl Sender {
    pub async fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
        let sender_wallet = get_client(Some("sender"));
        let amount_to_send = pj_uri.amount.unwrap();
        let receiver_address = pj_uri.address.clone();
    }
}
```

* Construct an "Original PSBT" paying `.amount` to `.address` as specified in
  the URI.

  First we add the reciever address to the outputs list along side the request
  amount. Then we create a funded psbt with options set to specify `fee_rate`

>   and other paramters.

Reviewer: explain why we're making an original psbt and how it's just a fallback in case we disappear

Author: sorry, can u please explain further? which side would disappear?

```rust
...
pub async fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {

>     ...

Reviewer: comment out `...` so it's not so easy to copy an error

    let mut outputs = HashMap::with_capacity(1);
    outputs.insert(receiver_address.to_string(), amount_to_send);
    let options = bitcoincore_rpc::json::WalletCreateFundedPsbtOptions {
        lock_unspent: Some(false),
        fee_rate: Some(bitcoincore_rpc::bitcoin::Amount::from_sat(10000)),
        ..Default::default()
    };
    let sender_psbt = sender_wallet
        .wallet_create_funded_psbt(
            &[], // inputs
            &outputs,
            None, // locktime
            Some(options),
            None,
        )
        .unwrap();
    let psbt = sender_wallet
        .wallet_process_psbt(&sender_psbt.psbt, None, None, None)
        .unwrap()
        .psbt;
    let psbt = Psbt::from_str(&psbt).unwrap();
}
```

* Construct the HTTPS request

  Prepare the request we want to send and also get the context back so we can

>   handle the response when its recieved.

Reviewer: This is a good opportunity to explain why authenticated communication is necessary and plaintext is insufficient.

        .unwrap();
    let res = res.text().await.unwrap();
}
```

* Process the response
  
  Validate PSBT inputs, outputs and fees as it may have been altered by the
  reciever.

```rust
...
pub async fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    ...
    let psbt = ctx.process_response(&mut res.as_bytes()).unwrap();

}
```

* Sign and finalize the Payjoin Proposal PSBT

  In the previous step we validated the PSBT and now its time finish the
  transaction, sign the PSBT and broadcast it.

```rust
...
pub async fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    ...
    let psbt = sender_wallet
        .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, None)
        .unwrap()
        .psbt;
    let tx = sender_wallet
        .finalize_psbt(&psbt, Some(true))
        .unwrap()
        .hex
        .unwrap();
    let txid = sender_wallet.send_raw_transaction(&tx).unwrap();
    txid
}
```
* And lets call the sender functions in the `main`.

  In the code below, `args` also read from the command line to consume the URI
  and utilise `payjoin::Uri::try_from` to convert the string to a valid URI.

```rust

> #[tokio::main]

Reviewer: You can use `reqwest::blocking::Client` to avoid the complexity of an async runtime with reqwest's `blocking` feature

        .wallet_create_funded_psbt(
                &[], // inputs
                &outputs,
                None, // locktime
                Some(options),
                None,
                )
        .unwrap();
    let psbt = sender_wallet
        .wallet_process_psbt(&sender_psbt.psbt, None, None, None)
        .unwrap()
        .psbt;
    let psbt = Psbt::from_str(&psbt).unwrap();
// }
```

##### 3.5 Construct the HTTPS request

Prepare the request we want to send and also get the context back so we can
handle the response when its received.

```rust
// ...
// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    // ...
        let (req, ctx) = RequestBuilder::from_psbt_and_uri(psbt, pj_uri)
        .unwrap()
        .build_with_additional_fee(
                bitcoincore_rpc::bitcoin::Amount::from_sat(1),
                None,
                bitcoincore_rpc::bitcoin::FeeRate::MIN,
                true,
                )
        .unwrap()
        .extract_v1()
        .unwrap();
// }
```

##### 3.6 Send the request 

```rust
// ...
// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    // ...

>         let res = https_agent()

Reviewer: ```suggestion
    let res = https_agent()
```

nit fmt

>         .post(req.url.as_str()) // pj_uri.endpoint
>         .header("Content-Type", "text/plain")
>         .body(String::from_utf8(req.body).unwrap())

Reviewer: ```suggestion
        .post(req.url) // pj_uri.endpoint
        .header("Content-Type", "text/plain")
        .body(req.body)
```
The dev kit intends to reduce the need for any HTTP I/O adapter. It should give you appropriate types to pass directly to an HTTP client. If not that might be a bug we can fix.

Author: I accepted this change without testing it. and then I saw your comment. why are you suggesting a code change, with the code you want to see without testing it first?

I spent sometime testing each step and made sure its working. I do expect, if you suggest code changes, to at least validate them locally first?

        .send()
        .unwrap();
    let res = res.text().unwrap();
// }
```

##### 3.7 Process the response

Validate PSBT inputs, outputs and fees as it may have been altered by
the receiver.

```rust
// ...
// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    // ...
    let psbt = ctx.process_response(&mut res.as_bytes()).unwrap();

// }
```

##### 3.8 Sign and finalize the Payjoin Proposal PSBT

> In the previous step we validated the PSBT and now its time finish the
> transaction, sign the PSBT and broadcast it.

Reviewer: ```suggestion
In the previous step we validated the PSBT and now it's time finish the
transaction, sign the PSBT and broadcast it.
```

```rust
// ...
// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    // ...
        let psbt = sender_wallet

>         .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, None)

Reviewer: ```suggestion
        .wallet_process_psbt(&psbt.to_string(), None, None, None)
```

`Psbt::to_string()` serializes to base64 encoding

>         let psbt = sender_wallet
>         .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, None)
>         .unwrap()
>         .psbt;
>     let tx = sender_wallet
>         .finalize_psbt(&psbt, Some(true))
>         .unwrap()
>         .hex
>         .unwrap();
>     let txid = sender_wallet.send_raw_transaction(&tx).unwrap();
>     txid

Reviewer: ```suggestion
    let psbt = sender_wallet
        .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, None)
        .unwrap()
        .psbt;
    let tx = sender_wallet
        .finalize_psbt(&psbt, Some(true))
        .unwrap()
        .hex
        .unwrap();
    let txid = sender_wallet.send_raw_transaction(&tx).unwrap();
    txid
```

nit fmt

// }
```

##### 3.9 And lets call the sender functions in the `main`.

> In the code below, `args` also read from the command line to consume the URI
> and utilise `payjoin::Uri::try_from` to convert the string to a valid URI.

Reviewer: ```suggestion
In the code below, `args` also reads from the command line to consume the URI
and utilizes `payjoin::Uri::try_from` to convert the string to a valid URI.
```

conjugation, spelling

I also notice a switch between URI and Uri. Is that arbitrary, or specific to the spec name vs rust type? If it's specific one could use "URI" and "`Uri`" to make the difference consistent.

```rust
fn main() {
    let args: Vec<String> = std::env::args().collect();
    let payjoin_uri: String = args[1].parse().unwrap();
    let payjoin_uri = payjoin::Uri::try_from(payjoin_uri)
        .unwrap()
        .assume_checked();
    let txid = sender::try_payjoin(payjoin_uri);
    dbg!(&txid);
}

> ```

Reviewer: I accidentally included this in the sender module when following along. Perhaps say "Back at the top of the file, replace the existing `main()` function or similar. Easy to fix and somewhat obvious by your lack of inline comments in the rust block, but it was my experience following along.

Author: I actually removed the whole file before I started doing the tutorial, so no "back to top" really. I think everyone does this a bit differently. anyhow, I added a comment to emphasis that we are back to the main function

## 4. Receiver

To setup the receiver, we will use `payjoin-cli` a cli payjoin implementation
developed by the `rust-payjoin` team. We will setup this receiver, generate
BiP21 URI and go back to the `sender` we implemented and give it the URI.

Open your terminal, and:

#### 4.1 Install `payjoin-cli`

```shell
cargo install payjoin-cli --version 0.0.4-alpha --features danger-local-https
```

#### 4.2 Run `payjoin-cli` as the receiver

```rust
payjoin-cli -r http://localhost:18443/wallet/receiver --rpcuser foo --rpcpass qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0= receive --endpoint https://localhost:3000 1230000000
```

Once you run the the above command, an https server will start and a
payjoin URI will be printed, and will look like this:

```shell
bitcoin:bcrt1qstujjzput7eznwv27e7qh7lpkgz33fh39g2gpt?amount=12.3&pj=https://localhost:3000/&pjos=0
```

> Copy the payjoin uri as we will need it in the next(final) step.

Reviewer: ```suggestion
Copy the payjoin uri as we will need it in the next (final) step.
```

I'm familiar with a space outside parens in a written clause being convention. Obviously not in maths/science function calls.
https://www.scribens.com/grammar-rules/typography/spacing.html

## 5. Run

Lets navigate back to the `payjoin-tutorial` folder and let the sender make a
request to the payjoin uri generated by the receiver in the previous step.

Replace `payjoin-uri` with the uri you copied in the previous step.

```shell
cargo run -- "payjoin-uri"
```

```shell
### cargo run -- "bitcoin:bcrt1qstujjzput7eznwv27e7qh7lpkgz33fh39g2gpt?amount=12.3&pj=https://localhost:3000/&pjos=0"

> ```

Reviewer: I think your instructions here are clear and I appreciate the two code blocks to make the example.

It would also be really cool if either the payjoin tx were pretty printed with an explanation of the inputs/outputs coming from sender and receiver, or if there were instructions on how to view it from regtest cli.

Author: I think we can leave that to a whole separate tutorial. explaining what we do in the inputs and how that is different from normal tx.

## Conclusion

We just demonstrated how to integrate `rust-payjoin` with a bitcoin

> wallet, using only the `sender` functionality. You could enable other

Reviewer: ```suggestion
wallet, using only the [`send` feature module](https://docs.rs/payjoin/latest/payjoin/send/index.html). You could enable other
```

This is a good chance to make it easy to follow a link for those ready to learn more

> features like `receiver` and `v2` to build more functionality.  We

Reviewer: ```suggestion
features like [`receive`](https://docs.rs/payjoin/latest/payjoin/receive/index.html) and `v2` to build more functionality.  We
```

there's no `receiver` feature ;)

also used the `payjoin-cli` to perform as a receiver, while we could

> also utilise it to be a sender.

Reviewer: ```suggestion
also utilize it to be a sender.
```

> Thanks for following along.

Reviewer: Ideally this article acts as a funnel to get people to use Payjoin Dev Kit in their projects. Consider starting this section with "Thanks for following along." and concluding with "... when you're ready to do more with payjoin check out x and y modules in PDK or using payjoin-cli as a sender" with links.

            .header("Content-Type", "text/plain")
            .body(String::from_utf8(req.body).unwrap())
            .send()
            .await
            .unwrap();
        let res = res.text().await.unwrap();
        // Step 6. Process the response
        //
        // An `Ok` response should include a Payjoin Proposal PSBT.
        // Check that it's signed, following protocol, not trying to steal or otherwise error.
        let psbt = ctx.process_response(&mut res.as_bytes()).unwrap();
        // Step 7. Sign and finalize the Payjoin Proposal PSBT
        //
        // Most software can handle adding the last signatures to a PSBT without issue.
        let psbt = sender_wallet
            .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, None)
            .unwrap()
            .psbt;
        let tx = sender_wallet
            .finalize_psbt(&psbt, Some(true))
            .unwrap()
            .hex
            .unwrap();
        // Step 8. Broadcast the Payjoin Transaction
        let txid = sender_wallet.send_raw_transaction(&tx).unwrap();
        txid
    }
}

```

And lets call the sender functions in the main function:

> ```

Reviewer: ```rust

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    setup_wallets();
    let args: Vec<String> = std::env::args().collect();
    let payjoin_uri: String = args[1].parse().unwrap();
    let payjoin_uri = payjoin::Uri::try_from(payjoin_uri)
        .unwrap()
        .assume_checked();
    let txid = Sender::try_payjoin(payjoin_uri).await;
    dbg!(&txid);
    Ok(())
}
```

# 5. Receiver
 To setup the receiver, we will use `payjoin-cli` a cli payjoin
 implementation developed by the `rust-payjoin` team. We will setup
 this receiver, generate BiP21 URI and go back to the `sender` we
 implemented and give it the URI.

 Open your terminal, and:

 1. Install `payjoin-cli`

 ```shell

>  cargo install payjoin-cli --version 0.0.2-alpha --features local-danger-https

Reviewer: /local-danger-https/danger-local-https/

This is failing due to rustls version mismatch in the danger-local-https deps. We'll have to release a fix this before the tutorial ships. I'll proceed using a locally built payjoin-cli.

 ```

 2. Run `payjoin-cli` as the receiver

 *Here we assume that you completed all of the previous steps
 successfully and generated the `receiver` wallet*.

 ```rust

>   payjoin-cli -r http://localhost:18443/wallet/reciever --rpcuser foo --rpcpass qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0= receive --endpoint https://localhost:3000 1230000000

Reviewer: A few things could simplify this manual setup

    Writing a script that automatically runs a payjoin-cli receiver in the docker container
    using a simple password like "insecurepassword"
    using the default cookie auth and having the container output the cookie location on startup

Looks like the script also created a wallet called receiver while this command wants a wallet called reciever (the former is common english)

remove the indentation

using Rust.  We'll delve into the `payjoin` crate, providing
step-by-step instructions on crafting and sending your very own
Payjoin transactions. Moreover, We'll guide you through setting up
payjoin-cli, and use it as the receiver side in the payjoin
transaction.

The `payjoin` crate provides an API to construct a Payjoin Proposal
PSBT, send it to the receiver and then process the response. The
sender can then sign and finalize the transaction and broadcast it to
the bitcoin network.

While this tutorial centers on the sender perspective, remember that
the payjoin crate also empowers you to act as a receiver, providing
greater flexibility. We encourage you to delve into its full potential
to unlock the comprehensive Payjoin experience.

By following this hands-on guide, you'll gain the knowledge and skills
to harness the power of Payjoin for your Bitcoin transactions,
enhancing both privacy and potentially reducing fees. Are you ready to
embark on this journey? Let's begin!

## 1. Bitcoin Node & Wallets

In order to perform a payjoin transaction, you need to have two wallets. 
lets use [regtest-util](https://github.com/jbesraa/regtest-util) to

> setup the bitcoin wallets.

Reviewer: ```suggestion
set up the bitcoin wallets.
```
```suggestion
setup the bitcoin wallets.
```

**You need to have Docker running and Cargo installed**

```shell
cargo install regtest-util
```

Be sure to add the `.cargo/bin` folder to your PATH to be able to run
the installed binaries.

```shell
regtest-util
```

After running the above commands, you will have two wallets, `sender`
and `receiver` running on `localhost:18443/sender` and `localhost:18443/receiver`
respectively.

## 2. Project Setup

Lets set up a new rust project and name it `payjoin-tutorial`

```shell
cargo new payjoin-tutorial
```

Add required dependencies to `Cargo.toml` so it includes the
following:

```toml
[dependencies]
bitcoincore-rpc = "0.17.0" 
payjoin = { version = "0.13.0", features = ["send", "base64"]  }
reqwest = {version = "0.11.24", features = ["rustls-tls", "blocking"]}
tokio = {version = "1.36.0", features = ["full"]}
```

## 3. Sender

In order to send a payjoin transaction, the sender needs to complete two steps.

> First, the sender obtains the receiver [BIP21 Bitcoin URI](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki) (often via QR code) and construct a PSBT

Reviewer: ```suggestion
First, the sender obtains the receiver [BIP21 Bitcoin URI](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki) (often via QR code) and constructs a PSBT
```

        .header("Content-Type", "text/plain")
        .body(req.body)
        .send()
        .unwrap();
    let res = res.text().unwrap();
// }
```

##### 3.7 Process the response

Validate PSBT inputs, outputs and fees as it may have been altered by
the receiver.

```rust
// ...
// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    // ...
    let psbt = ctx.process_response(&mut res.as_bytes()).unwrap();

// }
```

##### 3.8 Sign and finalize the Payjoin Proposal PSBT

In the previous step we validated the PSBT and now it's time finish
the transaction, sign the PSBT and broadcast it.

```rust
// ...
// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    // ...
        let psbt = sender_wallet
        .wallet_process_psbt(&psbt.to_string(), None, None, None)
        .unwrap()
        .psbt;
    let tx = sender_wallet
        .finalize_psbt(&psbt, Some(true))
        .unwrap()
        .hex
        .unwrap();
    let txid = sender_wallet.send_raw_transaction(&tx).unwrap();
    txid
// }
```

##### 3.9 And lets call the sender functions in the `main`.

Back to the `main` function, we will read the payjoin uri from the

> command line arguments and utilize `payjoin::Uri` to parse it.

Reviewer: ```suggestion
command line arguments and utilize `payjoin::Uri` to parse them.
```

parse [the arguments]

sender can then sign and finalize the transaction and broadcast it to
the bitcoin network.

While this tutorial centers on the sender perspective, remember that
the payjoin crate also empowers you to act as a receiver, providing
greater flexibility. We encourage you to delve into its full potential
to unlock the comprehensive Payjoin experience.

By following this hands-on guide, you'll gain the knowledge and skills
to harness the power of Payjoin for your Bitcoin transactions,
enhancing both privacy and potentially reducing fees. Are you ready to
embark on this journey? Let's begin!

## 1. Bitcoin Node & Wallets

In order to perform a payjoin transaction, you need to have two wallets. 
lets use [regtest-util](https://github.com/jbesraa/regtest-util) to
setup the bitcoin wallets.

**You need to have Docker running and Cargo installed**

```shell
cargo install regtest-util
```

Be sure to add the `.cargo/bin` folder to your PATH to be able to run
the installed binaries.

```shell
regtest-util
```

After running the above commands, you will have two wallets, `sender`
and `receiver` running on `localhost:18443/sender` and `localhost:18443/receiver`
respectively.

## 2. Project Setup

> Lets set up a new rust project and name it `payjoin-tutorial`
> ```shell

Reviewer: ```suggestion
Lets set up a new rust project and name it `payjoin-tutorial`

```shell
```

nit

Write a utility function to get the sender wallet, using `bitcoincore_rpc` to
connect to the wallet. The `Auth` struct is used to authenticate the
connection, and the `Client` struct is used to connect to the wallet.

```rust
// payjoin--tutorial/src/main.rs
// mod sender {
    fn get_sender_wallet() -> bitcoincore_rpc::Client {
        let auth = bitcoincore_rpc::Auth::UserPass(
            "foo".to_string(),
            "qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0=".to_string(),
        );
        bitcoincore_rpc::Client::new("http://localhost:18443/wallet/sender", auth).unwrap()
    }
    // ..
// }
```

##### 3.3 Implement `https_agent` function inside `sender` module.

The `https_agent` function is used in order to make https requests.
The sender will use this function to send the payjoin request to the
receiver.

Here, we are using `danger_accept_invalid_certs` as we run this
locally. In production this should be initiated with valid
certificates only.

[Why to use HTTPS in production](https://www.cloudflare.com/learning/ssl/why-is-http-not-secure/)

For Payjoin, it is important to use HTTPS to ensure the privacy of the

> transaction and to prevent potential lose of funds.  Without

Reviewer: ```suggestion
transaction and to prevent potential loss of funds.  Without
```

        let https = reqwest::blocking::Client::builder()
            .danger_accept_invalid_certs(true)
            .build()
            .expect("failed to build https client");
        https
    }
    // ..
// }
```

In the next few steps, we will add the core implementation for the
sender inside the `try_payjoin` function.

##### 3.4 Extract the parameters from the payjoin URI

Payjoin requires messages to be sent between sender and receiver. The
receiver hosts an HTTP endpoint to receive these messages. To make a
request, we need the receiver payjoin Uri. 

The Uri contains the amount the receiver is requesting and the address
they wish to receive to. Moreover, the Uri also includes the endpoint
we will make the request to, and that is extracted in the
`RequestContext::from_psbt_and_uri`.

```rust
// mod sender {
    // ..
    pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
        let sender_wallet = get_sender_wallet();
        let amount_to_send = pj_uri.amount.unwrap();
        let receiver_address = pj_uri.address.clone();
    }
// }
```

##### 3.4 Construct an "Original PSBT" paying `.amount` to `.address`

> as specified in the URI.

Reviewer: I know I suggested that a space be added after headings, but perhaps this line is this actually meant to be part of the heading?

It seems you're using automatic newlines, and I don't believe markdown will render headings on multiple lines. When I write markdown I just have my editor wordwrap and write long lines.

Author: yea it was a mistaken line-fixing. 

removed totally the .amount and .address from header as they are mentioned in the body just next to it.

In the previous step we validated the PSBT and now it's time finish
the transaction, sign the PSBT and broadcast it.

```rust
// ...
// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
    // ...
        let psbt = sender_wallet
        .wallet_process_psbt(&psbt.to_string(), None, None, None)
        .unwrap()
        .psbt;
    let tx = sender_wallet
        .finalize_psbt(&psbt, Some(true))
        .unwrap()
        .hex
        .unwrap();
    let txid = sender_wallet.send_raw_transaction(&tx).unwrap();
    txid
// }
```

##### 3.9 And lets call the sender functions in the `main`.

Back to the `main` function, we will read the payjoin uri from the
command line arguments and utilize `payjoin::Uri` to parse it.
After that, we will call the `try_payjoin` function and print the
transaction id.

```rust
fn main() {
    let args: Vec<String> = std::env::args().collect();
    let payjoin_uri: String = args[1].parse().unwrap();
    let payjoin_uri = payjoin::Uri::try_from(payjoin_uri)
        .unwrap()
        .assume_checked();
    let txid = sender::try_payjoin(payjoin_uri);
    dbg!(&txid);
}
```

## 4. Receiver

> To setup the receiver, we will use `payjoin-cli` a cli payjoin
> implementation developed by the `rust-payjoin` team. We will setup
> this receiver, generate BiP21 URI and go back to the `sender` we
> implemented and give it the URI.

Reviewer: ```suggestion
To setu p the receiver, we will use `payjoin-cli` a cli payjoin
implementation developed by the `rust-payjoin` team. We will setup
this receiver, generate BIP21 URI and go back to the `sender` we
implemented and give it the URI.
```

> To setup the receiver, we will use `payjoin-cli` a cli payjoin
> implementation developed by the `rust-payjoin` team. We will setup
> this receiver, generate BiP21 URI and go back to the `sender` we
> implemented and give it the URI.
> 
> Open your terminal, and:

Reviewer: ```suggestion
To set up the receiver, we will use `payjoin-cli` a cli payjoin
implementation developed by the `rust-payjoin` team. We will setup
this receiver, generate BIP21 URI and go back to the `sender` we
implemented and give it the URI.

Open your terminal, and:
```

# Introduction

This tutorial goal is to get you started with payjoin and demonstrate
payjoin capabilities. We will demonstrate how to setup the sender side
in a payjoin transaction and transact with a payjoin receiver
utilising `payjoin-cli`.

> While `rust-payjoin` supports both BiP0078 and BiP0077, this tutorial

Reviewer: Any time a spec is referenced it's got to be hyperlinked

is specificly to demonstrate BiP0078(payjoin version 1).

For a quick wallet setup, we utilise `docker` and `bitcoincore_rpc`
crate to run a regtest node and fund two wallets, one for the sender
and one for the receiver.

For any feedback, please free to reach out through the payjoin

> official discord or github.

Reviewer: Any call to action should also be completed with hyperlinks. Don't make the reader do more work than strictly necessary. This should be a welcoming tutorial after all

# 1. Bitcoin Node
 The following command will start a bitcoin core node in regtest mode,
 with auth setup already in place just for testing purposes.

 *You would need to have docker installed in order to run the
 following command*

 ```
 docker run --name bitcoin-server -d --rm -it \
 -p 18443:18443 \
 -p 18444:18444 \
 ruimarinho/bitcoin-core \
 -printtoconsole \
 -regtest=1 \
 -rpcallowip=0.0.0.0/0 \
 -rpcbind=0.0.0.0 \
 -rpcauth='foo:7d9ba5ae63c3d4dc30583ff4fe65a67e$9e3634e81c11659e3de036d0bf88f89cd169c1039e6e09607562d54765c649cc'

>  ```

Reviewer: Docker might be able to reckon with to write a tutorial, but it's more work for the person following the tutorial and leaves side-effects on their system. I'd much rather we use the [`bitcoind` crate](https://docs.rs/bitcoind/latest/bitcoind/) to handle that setup and tear down for them.

fn setup_wallets() {
    let client = bitcoincore_rpc::Client::new(
        "http://localhost:18443",
        Auth::UserPass(
            "foo".to_string(),
            "qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0=".to_string(),
        ),
    )
    .unwrap();
    let wallets = client.list_wallets().unwrap();
    dbg!(&wallets);
    if !wallets.contains(&"sender".to_string()) {
        dbg!(
            "creating sender
    wallet"
        );
        assert!(client
            .create_wallet("sender", None, None, None, None)
            .is_ok());
    }
    if !wallets.contains(&"receiver".to_string()) {
        dbg!("creating receiver wallet");
        assert!(client
            .create_wallet("receiver", None, None, None, None)
            .is_ok());
    }

    let sender_client = bitcoincore_rpc::Client::new(
        "http://localhost:18443/wallet/sender",
        Auth::UserPass(
            "foo".to_string(),
            "qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0=".to_string(),
        ),
    )
    .unwrap();
    let receiver_client = bitcoincore_rpc::Client::new(
        "http://localhost:18443/wallet/receiver",
        Auth::UserPass(
            "foo".to_string(),

>             "qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0=".to_string(),

Reviewer: Auth and the url should be saved in variables so they don't need to be repeated

>         ),

Reviewer: Can auth be saved in a variable so as not to repeat it?

  to convert the string to a valid URI.

  lets add the code to `main.rs` as well, just after the
  `setup_wallet` function:

  ```rust
// payjoin-tutorial/src/main.rs

use payjoin::bitcoin::address::NetworkChecked;
use std::collections::HashMap;
use std::path::PathBuf;
use std::str::FromStr;
use std::sync::Arc;

use bitcoincore_rpc::bitcoin::psbt::Psbt;
use bitcoincore_rpc::bitcoin::Txid;
use payjoin::base64;
use payjoin::send::RequestBuilder;
use rustls::client::ClientConfig;
use rustls::{Certificate, RootCertStore};
  
const LOCAL_CERT_FILE: &str = "localhost.der";

fn https_agent() -> ureq::Agent {
    use ureq::AgentBuilder;

    let mut local_cert_path = PathBuf::from("/tmp");
    local_cert_path.push(LOCAL_CERT_FILE);
    let cert_der = std::fs::read(local_cert_path).unwrap();
    let mut root_cert_store = RootCertStore::empty();
    root_cert_store.add(&Certificate(cert_der)).unwrap();
    let client_config = ClientConfig::builder()
        .with_safe_defaults()
        .with_root_certificates(root_cert_store)
        .with_no_client_auth();

    AgentBuilder::new()
        .tls_config(Arc::new(client_config))
        .build()

> }

Reviewer: Payjoin Dev Kit uses ureq in order to support msrv, but it's a less familiar crate than `reqwest`. One should be able to use a `reqwest::blocking::Client` with [`danger_accept_invalid_certs`](https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html#method.danger_accept_invalid_certs) to demonstrate this example without reading from a local cert in a file every run
